python/pyspark/sql/dataframe.py

```diff
@@ -2200,10 +2200,13 @@ def _collectAsArrow(self):
         .. note:: Experimental.
         """
         with SCCallSiteSync(self._sc) as css:
-            sock_info = self._jdf.collectAsArrowToPython()
+            port, auth_secret, jserver_obj = self._jdf.collectAsArrowToPython()
 
         # Collect list of un-ordered batches where last element is a list of correct order indices
-        results = list(_load_from_socket(sock_info, ArrowCollectSerializer()))
+        from pyspark.rdd import _create_local_socket
+        sock_file = _create_local_socket((port, auth_secret))
+        results = list(ArrowCollectSerializer().load_stream(sock_file))
+        jserver_obj.getResult()  # Join serving thread and raise any exceptions
```

Comment:
we might want to have this in a `finally` clause, so that if we have an error during serialization (which might be caused by an exception in the JVM) we will still get the original exception.